token/token_verify.go

```diff
@@ -267,53 +195,47 @@ func CheckAnyAuth(ctx *gin.Context, authOption AuthOption) bool {
 		case Authz:
 			authzToken := ctx.Request.URL.Query()["authz"]
 			if len(authzToken) <= 0 {
-				errMsg += fmt.Sprintln("No Authz query parameter present")
 				continue
 			} else {
 				token = authzToken[0]
 				tokenFound = true
 				break
 			}
 		default:
-			log.Info("Authentication failed. Invalid/unsupported token source")
-			return false
+			log.Error("Invalid/unsupported token source")
+			return http.StatusInternalServerError, false, errors.New("Cannot verify token due to bad server configuration. Invalid/unsupported token source")
 		}
 	}
 
 	if token == "" {
-		log.Info("Authentication failed. No token is present from the list of potential token positions")
-		return false
+		log.Debug("Unauthorized. No token is present from the list of potential token positions")
+		return http.StatusUnauthorized, false, errors.New("Authentication is required but no token is present.")
 	}
 
+	errMsg := ""
 	for _, iss := range authOption.Issuers {
 		switch iss {
 		case Federation:
-			err := authChecker.FederationCheck(ctx, token, authOption.Scopes, authOption.AllScopes)
-			if _, exists := ctx.Get("User"); err != nil || !exists {
-				errMsg += fmt.Sprintln("Token validation failed with federation issuer: ", err)
-				log.Debug("Token validation failed with federation issuer: ", err)
+			if err := authChecker.federationCheck(ctx, token, authOption.Scopes, authOption.AllScopes); err != nil {
+				errMsg += fmt.Sprintln("Cannot verify token with federation issuer: ", err)
 				break
 			} else {
-				log.Debug("Token validation succeeded with federation issuer")
-				return exists
+				return http.StatusOK, true, nil
 			}
 		case Issuer:
-			err := authChecker.IssuerCheck(ctx, token, authOption.Scopes, authOption.AllScopes)
-			if _, exists := ctx.Get("User"); err != nil || !exists {
-				errMsg += fmt.Sprintln("Token validation failed with server issuer: ", err)
-				log.Debug("Token validation failed with server issuer: ", err)
+			if err := authChecker.issuerCheck(ctx, token, authOption.Scopes, authOption.AllScopes); err != nil {
+				errMsg += fmt.Sprintln("Cannot verify token with server issuer: ", err)
 				break
 			} else {
-				log.Debug("Token validation succeeded with server issuer")
-				return exists
+				return http.StatusOK, true, nil
 			}
 		default:
-			log.Info("Authentication failed. Invalid/unsupported token issuer")
-			return false
+			log.Error("Invalid/unsupported token issuer")
+			return http.StatusInternalServerError, false, errors.New("Cannot verify token due to bad server configuration. Invalid/unsupported token issuer")
```

Comment:
If it's a bad token issuer, should this be a 403?

Reply:
The token issuer was set by us when evoking the verify function. Although a very rare situation, this is saying that somebody is not picking the issuer from the defined constants, so it's an internal error

Reply:
Cool, thanks for the explanation.